### PR TITLE
[join-lateral] 

### PR DESCRIPTION
**Pull request by rex-remind101**
added support for `JOIN LATERAL` as specified in MySQL and Postgres https://github.com/SeaQL/sea-query/issues/201

### CODE REVIEW

File: src/query/select.rs [hunk 1]
```diff
@@ -1061,14 +1064,86 @@ impl SelectStatement {
...
+    ///     r#"SELECT `name` FROM `font` LEFT JOIN LATERAL (SELECT `id` FROM `glyph`) AS `sub_glyph` ON `font`.`id` = `sub_glyph`.`id`"#
+    /// );
+    /// assert_eq!(
+    ///     query.to_string(PostgresQueryBuilder),
+    ///     r#"SELECT "name" FROM "font" LEFT JOIN LATERAL (SELECT "id" FROM "glyph") AS "sub_glyph" ON "font"."id" = "sub_glyph"."id""#
+    /// );
+    /// assert_eq!(
+    ///     query.to_string(SqliteQueryBuilder),
+    ///     r#"SELECT `name` FROM `font` LEFT JOIN LATERAL (SELECT `id` FROM `glyph`) AS `sub_glyph` ON `font`.`id` = `sub_glyph`.`id`"#
+    /// );
```
Comment (billy1624) [inline src/query/select.rs:1101]: JOIN `LATERAL` seems not supported by SQLite. We shouldn't generate invalid select statement

File: src/query/select.rs [hunk 2]
```diff
@@ -1061,14 +1064,86 @@ impl SelectStatement {
...
+    pub fn join_lateral_subquery<T, C>(
+        &mut self,
+        join: JoinType,
+        query: SelectStatement,
+        alias: T,
+        condition: C,
+    ) -> &mut Self
+    where
+        T: IntoIden,
+        C: IntoCondition,
```
Comment (billy1624) [inline src/query/select.rs:1129]: I'm thinking this method could simply be named `join_lateral`. As it's assumed to take a (sub)query as argument. What do you think?

File: src/query/select.rs (hunk 1 above)
Comment (rex-remind101) [reply src/query/select.rs:1101]: how do i error on sqlite within the builder?

File: src/query/select.rs (hunk 2 above)
Comment (rex-remind101) [reply src/query/select.rs:1129]: yup, that makes sense 👍

File: src/query/select.rs (hunk 1 above)
Comment (billy1624) [reply src/query/select.rs:1101]: We could `panic!` when building the select statement for SQLite. Is this be reasonable? @tyt2y3 It's better to let the user know explicitly something went wrong early.

**Comment by tyt2y3** [comment]
I think the `lateral` should be made part of `JoinType` though.

**Comment by rex-remind101** [comment]
@billy1624 hello, any thoughts?

**Comment by billy1624** [comment]
> I noticed this was merged into it's own branch. Has any further work been done on it? Should I make another pr with those changes or is there a different direction we're heading in?

It was later cherry picked into master, see https://github.com/SeaQL/sea-query/commit/12c1292c2e890c06527d9a2a3fd7b3a4d8a24508

**Comment by rex-remind101** [comment]
thank you, do i need to follow up with the above adjustments you suggested? if so, i'm still not sure where to insert the `panic!` for slqite, thanks

**Comment by billy1624** [comment]
> thank you, do i need to follow up with the above adjustments you suggested? if so, i'm still not sure where to insert the `panic!` for slqite, thanks

I can open a PR for that :P

**Comment by billy1624** [comment]
> thank you, do i need to follow up with the above adjustments you suggested? if so, i'm still not sure where to insert the `panic!` for slqite, thanks

Hey @rex-remind101, I have open an PR (#243) for this

**Pull request merged**